geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/FilterHelper.scala

```diff
@@ -418,6 +430,54 @@ object FilterHelper {
     Some(FilterValues(Seq(Bounds.everything[T]), precise = false))
   }
 
+  /**
+   * Calculates all the different case permutations of a string.
+   *
+   * For example, "foo" -> Seq("foo", "Foo", "fOo", "foO", "fOO", "FoO", "FOo", "FOO")
+   *
+   * @param string input string
+   * @return
+   */
+  private def casePermutations(string: String): Seq[String] = {
+    val max = FilterProperties.CaseInsensitiveLimit.toInt.getOrElse {
+      // has a valid default value so should never return a none
+      throw new IllegalStateException(
+        s"Error getting default value for ${FilterProperties.CaseInsensitiveLimit.property}")
+    }
+
+    val lower = string.toLowerCase(Locale.US)
+    val upper = string.toUpperCase(Locale.US)
+    // account for chars without upper/lower cases, which we don't need to permute
+    val count = (0 until lower.length).count(i => lower(i) != upper(i))
+
+    if (count > max) {
+      FilterHelperLogger.log.warn(s"Not expanding case-insensitive prefix due to length: $string")
+      Seq.empty
+    } else {
+      // there will be 2^n different permutations, accounting for chars that don't have an upper/lower case
+      val permutations = Array.fill(math.pow(2, count).toInt)(Array(lower: _*))
```

Comment:
I wonder if the call to 'fill' is 'expensive' here?  E.g., do we end up setting and resetting the values?

Reply:
we set them originally to the lowercase bytes (copying the array for each permutation). then we overwrite the uppercase ones in our loop. so the array values are being modified later but the arrays are just created here up front. alternatively we could instantiate empty arrays and then overwrite both upper and lower case chars in the loop, but i don't think that would be any better...